### PR TITLE
[db] BF-204 Migrate to MongoDB when save images

### PR DESCRIPTION
**Pull request by antonydanon**

### CODE REVIEW

File: src/main/java/by/bookify/image/controller/ImageController.java [hunk 1]
```diff
@@ -14,7 +16,7 @@
 @RequiredArgsConstructor
 public class ImageController {
 
-    private final ImageService imageService;
+    private final GridFsMongoImageService imageService;
```
Comment (korzdan) [inline src/main/java/by/bookify/image/controller/ImageController.java]: You have choosen the version explicitly. It's better to use contract in other words interfaces. Use Qualifier annotation. You can read about it here: https://www.baeldung.com/spring-qualifier-annotation

File: src/main/java/by/bookify/image/service/MongoImageService.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        Image image = imageRepository
+                .findById(bookId)
+                .orElseThrow(() -> new ImageNotFoundException("Image with " + bookId + " ID not found."));
+        return imageToInputStreamResource(image);
+    }
+
+    private Image buildImage(Long bookId, MultipartFile file) {
+        return Image.builder()
+                .id(bookId)
+                .title(bookId + JPG_IMAGE_EXTENSION)
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/MongoImageService.java]: JPG_IMAGE_EXTENSION delete please

File: src/main/java/by/bookify/image/service/MongoImageService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+@RequiredArgsConstructor
+@Slf4j
+public class MongoImageService implements ImageService {
+
+    private final ImageRepository imageRepository;
+
+    @Override
+    public void uploadImage(Long bookId, MultipartFile file) {
+        Image image = buildImage(bookId, file);
+        imageRepository.deleteById(bookId);
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/MongoImageService.java]: Check on null

File: src/main/java/by/bookify/image/controller/ImageController.java [hunk 4]
```diff
@@ -14,7 +16,7 @@
...
```
Comment (korzdan) [inline src/main/java/by/bookify/image/controller/ImageController.java:21]: And why there are so many realizations? May be you should choose one that suits best

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        DBObject metaData = createBasicDBObject(bookId);
+        gridFsTemplate.delete(new Query(Criteria.where("metadata.id").is(bookId)));
+        gridFsTemplate.store(fileToInputStream(file, bookId), file.getName(), file.getContentType(), metaData);
+    }
+
+    @Override
+    public InputStreamResource findImage(Long bookId) {
+        GridFSFile file = gridFsTemplate.findOne(new Query(Criteria.where("metadata.id").is(bookId)));
+        if(file == null)
+            throw new ImageNotFoundException("Image with " + bookId + " ID not found.");
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: Add curly braces and space

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+@AllArgsConstructor
+@Slf4j
+public class GridFsMongoImageService implements ImageService {
+
+    private GridFsTemplate gridFsTemplate;
+
+    @Override
+    public void uploadImage(Long bookId, MultipartFile file) {
+        DBObject metaData = createBasicDBObject(bookId);
+        gridFsTemplate.delete(new Query(Criteria.where("metadata.id").is(bookId)));
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: You have used metadata.id twice. Is it a constant value?

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        gridFsTemplate.store(fileToInputStream(file, bookId), file.getName(), file.getContentType(), metaData);
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: Extract to variable

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public InputStreamResource findImage(Long bookId) {
+        GridFSFile file = gridFsTemplate.findOne(new Query(Criteria.where("metadata.id").is(bookId)));
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: Extract to variable. And also try to write Query annotation in controller

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+@Service
+@AllArgsConstructor
+@Slf4j
+public class GridFsMongoImageService implements ImageService {
+
+    private GridFsTemplate gridFsTemplate;
+
+    @Override
+    public void uploadImage(Long bookId, MultipartFile file) {
```
Comment (korzdan) [inline src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: What if we delete file then create method throws exception?

**Review by korzdan — CHANGES_REQUESTED** [review]

File: src/main/java/by/bookify/image/service/MongoImageService.java (hunk 3 above)
Comment (KaShTaN27) [reply src/main/java/by/bookify/image/service/MongoImageService.java]: Should it be a transaction?

File: src/main/java/by/bookify/image/model/Image.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.data.mongodb.core.mapping.Document;
+
+@Data
+@Builder
+@Document(collection = "images")
+public class Image {
+    @Id
+    private Long id;
+
+    private String title;
```
Comment (KaShTaN27) [inline src/main/java/by/bookify/image/model/Image.java]: Does title really needed?

**Review by KaShTaN27 — CHANGES_REQUESTED** [review]

File: src/main/java/by/bookify/image/controller/ImageController.java (hunk 4 above)
Comment (antonydanon) [reply src/main/java/by/bookify/image/controller/ImageController.java:21]: if you want only one realization, remove it by yourself

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java (hunk 6 above)
Comment (antonydanon) [reply src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: okey man

File: src/main/java/by/bookify/image/service/GridFsMongoImageService.java (hunk 8 above)
Comment (antonydanon) [reply src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: In controller? May be you mean repository?

Comment (antonydanon) [reply src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: I tried - it's not good idea, in my opinion

Comment (korzdan) [reply src/main/java/by/bookify/image/service/GridFsMongoImageService.java]: Hahahah, it all was beer. I meant to write in "repository" not in a "controller"

**Review by korzdan — APPROVED** [review]

**Review by KaShTaN27 — APPROVED** [review]

**Pull request merged**